machinery-helper/machinery_helper.go

```diff
@@ -0,0 +1,231 @@
+      }
+    }
+  }
+
+  addImplicitlyManagedDirs(dirs, files)
+
+  return files, dirs
+}
+
+func assembleJson(unmanaged_files_map interface{}) string {
+  json_map := map[string]interface{}{ "extracted": false, "files": unmanaged_files_map }
+  json, _ := json.MarshalIndent(json_map, " ", "  ")
+  return string(json)
+}
+
+var readDir = func(dir string) ([]os.FileInfo, error) {
+  return ioutil.ReadDir(dir)
+}
+
+func hasManagedDirs(dir string, rpm_dirs map[string]bool) bool {
+  for rpm_dir := range rpm_dirs {
+    if strings.HasPrefix(rpm_dir, dir + "/") {
+       return true
+    }
+  }
+  return false
+}
+
+func findUnmanagedFiles(dir string, rpm_files map[string]string, rpm_dirs map[string]bool, unmanaged_files map[string]string) {
+  files, _ := readDir(dir)
+  for _, f := range files {
+    file_name := dir + f.Name()
+    if !utf8.ValidString(file_name) {
+      fmt.Fprintln(os.Stderr, file_name, "contains invalid UTF-8 characters. Skipping.")
+    } else {
+      if _, ok := Ignore_list[file_name]; !ok {
+        if f.IsDir() {
+          if _, ok := rpm_dirs[file_name]; ok {
+            findUnmanagedFiles(file_name + "/", rpm_files, rpm_dirs, unmanaged_files)
+          } else {
+            if !hasManagedDirs(file_name, rpm_dirs) {
+              unmanaged_files[file_name + "/"] = "dir"
+            }
+          }
+        } else {
+          if _, ok := rpm_files[file_name]; !ok {
+            if f.Mode() &
+              (os.ModeSocket | os.ModeNamedPipe | os.ModeDevice | os.ModeCharDevice) != 0 {
+              // Ignore sockets, named pipes and devices
+            } else if f.Mode() & os.ModeSymlink == os.ModeSymlink {
+              unmanaged_files[file_name] = "link"
+            } else {
+              unmanaged_files[file_name] = "file"
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+func print_version() {
+  fmt.Println("Version:", VERSION)
+  os.Exit(0)
+}
+
+func main() {
+  // check for tar extraction
+  if len(os.Args) >= 2 {
+    switch os.Args[1] {
+    case "tar":
+      Tar(os.Args[2:])
+      os.Exit(0)
+    }
+  }
+
+  // parse CLI arguments
+  var version_flag = flag.Bool("version", false, "shows the version number")
+  flag.Parse()
+
+  // show version
+  if (*version_flag == true) {
+    print_version()
+  }
+
+  // fetch unmanaged files
+  unmanaged_files := make(map[string]string)
+  this_binary, _ := filepath.Abs(os.Args[0])
+
+  ignore_list := map[string]bool{
+    this_binary: true,
+  }
+  for _, mount := range RemoteMounts() {
+    ignore_list[mount] = true
+  }
+  for _, mount := range SpecialMounts() {
+    ignore_list[mount] = true
+  }
+  Ignore_list = ignore_list
+
+  for _, mount := range RemoteMounts() {
+    unmanaged_files[mount + "/"] = "remote_dir"
+  }
+
+  rpm_files, rpm_dirs := getManagedFiles()
+  findUnmanagedFiles("/", rpm_files, rpm_dirs, unmanaged_files)
+
+  files := make([]string, len(unmanaged_files))
+  i := 0
+  for k, _ := range unmanaged_files {
```

Comment:
should omit 2nd value from range; this loop is equivalent to `for k := range ...`